middleware/kubernetes/nametemplate/nametemplate.go

```diff
@@ -0,0 +1,177 @@
+package nametemplate
+
+import (
+	"errors"
+    "fmt"
+    "strings"
+
+	"github.com/miekg/coredns/middleware/kubernetes/util"
+)
+
+// Likely symbols that require support:
+// ${id}
+// ${ip}
+// ${portname}
+// ${protocolname}
+// ${servicename}
+// ${namespace}
+// ${type}              "svc" or "pod"
+// ${zone}
+
+
+// SkyDNS normal services have an A-record of the form "${servicename}.${namespace}.${type}.${zone}"
+// This resolves to the cluster IP of the service.
+
+// SkyDNS headless services have an A-record of the form "${servicename}.${namespace}.${type}.${zone}"
+// This resolves to the set of IPs of the pods selected by the Service. Clients are expected to
+// consume the set or else use round-robin selection from the set.
+
+
+var symbols = map[string]string{
+    "service": "${service}",
+    "namespace": "${namespace}",
+    "type": "${type}",
+    "zone": "${zone}",
+}
+
+var types = []string{
+	"svc",
+	"pod",
+}
+
+
+// TODO: Validate that provided NameTemplate string only contains:
+//			* valid, known symbols, or
+//			* static strings
+
+// TODO: Support collapsing multiple segments into a symbol. Either:
+//			* all left-over segments are used as the "service" name, or
+//			* some scheme like "${namespace}.${namespace}" means use
+//			  segments concatenated with a "." for the namespace, or
+//			* ${namespace2:4} means use segements 2->4 for the namespace.
```

Comment:
does k8s need this stuff? Just wondering...

Also, as said, look into the log middleware which does similar paring and can probably be factored out and reused - We can do that in a subsequent PR though